File: medium-problem/0560_Subarray Sum Equals K.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */

// [brute force] -- TLE
var subarraySum = function(nums, k) {
    let count = 0;
    for(let end=0; end<nums.length; end++){
        let sum =0;
        for(let start=end; start>=0; start--){
            sum += nums[start]
            if(sum === k){
                count ++
            }
        }
    }
    return count
};
// TC: O(n^2)
// MC: O(1)

// [hash map]
var subarraySum = function(nums, k) {
  let map = new Map();
  let sum = 0;
  let count = 0;
  map.set(0,1);
  for (let i=0;i<nums.length;i++) {
      sum += nums[i];
      if (map.has(sum-k)) count += map.get(sum-k);
      if (map.has(sum)) map.set(sum, map.get(sum)+1);
      else map.set(sum, 1);
  }
  return count;
}
// TC: O(n)
// MC: O(n)

File: c#-solution/0560. Subarray Sum Equals K.cs
// 560. Subarray Sum Equals K
// https://leetcode.com/problems/subarray-sum-equals-k/
// Tag: C#, Array, Hash Table, Prefix Sum, Medium


// brute force approach
public class Solution {
    public int SubarraySum(int[] nums, int k) {
        var result =0;
        for(int i=0; i<nums.Length; i++){
            var sum = 0;
            for(int j=i; j<nums.Length; j++){
                sum += nums[j];
                if(sum == k) result ++;
            }
        }

        return result;
    }
}
// TC: O(n^2)
// SC: O(1)


// prefix sum + hash table approach
public class Solution {
    public int SubarraySum(int[] nums, int k) {
        var result = 0;
        var prefixSum = new Dictionary<int, int>();
        prefixSum[0] = 1; // Initialize with sum 0

        var currentSum = 0;
        for (int i = 0; i < nums.Length; i++) {
            currentSum += nums[i];
            if (prefixSum.ContainsKey(currentSum - k)) {
                result += prefixSum[currentSum - k];
            }
            if (prefixSum.ContainsKey(currentSum)) {
                prefixSum[currentSum]++;
            } else {
                prefixSum[currentSum] = 1;
            }
        }

        return result;
    }
}
// TC: O(n)
// SC: O(n)